docs/configuration.md
---
id: configuration
title: Configuration
sidebar_label: Configuration
---

The API server can be configured via the following environment variables:

| Environment variable         | Description                                                                                                                                                                            | Default value         |
| ---------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------- |
| NODE_ENV                     | The node runtime environment                                                                                                                                                           | dev                   |
| TR_PORT                      | The port on which the server will listen on                                                                                                                                            | 8080                  |
| TR_SECRET                    | The application secret, used i.e. for signing and verifying the auth tokens, it is **strongly** recommended that you do not use the default one                                        | secret                |
| TR_VIRTUAL_HOST              | The virtual host tells the server under which domain you are hosting the webapp, used i.e. to correctly format URLs and links in emails                                                | http://localhost:8080 |
| TR_DB_TYPE                   | The type of database to use                                                                                                                                                            | mysql                 |
| TR_DB_HOST                   | The DB host                                                                                                                                                                            | 127.0.0.1             |
| TR_DB_PORT                   | The DB port                                                                                                                                                                            | 3306                  |
| TR_DB_DATABASE               | The database name to connect to                                                                                                                                                        | tr_dev                |
| TR_DB_USER                   | The user to connect with                                                                                                                                                               | root                  |
| TR_DB_PASSWORD               | The password to connect with                                                                                                                                                           | (empty)               |
| TR_PUBLIC_DIR                | The path at which the public server files are located                                                                                                                                  | ./public              |
| TR_TEMPLATES_DIR             | The path at which the server templates are located                                                                                                                                     | ./templates           |
| TR_CORS_ENABLED              | If enabled, the server will allow CORS requests                                                                                                                                        | false                 |
| TR_ACCESS_LOGS_ENABLED       | If enabled, the server will log all requests to STDOUT                                                                                                                                 | true                  |
| TR_AUTH_TOKEN_EXPIRES        | The duration in seconds that the issued JWT tokens will be valid for                                                                                                                   | 86400                 |
| TR_SIGNUPS_ENABLED           | If enabled, users can sign in for                                                                                                                                                      | true                  |
| TR_MAX_PROJECTS_PER_USER     | The max number of projects that can be created per user                                                                                                                                | 100                   |
| TR_DEFAULT_PROJECT_PLAN      | The default project plan, by default set to _open-source_ and is limited to 100,000 strings (you are free to change it)                                                                | open-source           |
| TR_DB_AUTOMIGRATE            | If enabled, on startup the server will ensure that all DB migrations have been applied                                                                                                 | true                  |
| TR_MAIL_DEBUG                | If enabled, it log a preview email link                                                                                                                                                | false                 |
| TR_MAIL_SENDER               | The mail sender                                                                                                                                                                        | not set               |
| TR_MAIL_HOST                 | The mail server host                                                                                                                                                                   | not set               |
| TR_MAIL_PORT                 | The mail server port                                                                                                                                                                   | 587                   |
| TR_MAIL_SECURE               | If enabled, attempt to send mail securely                                                                                                                                              | false                 |
| TR_MAIL_REJECT_SELF_SIGNED   | If enabled, reject self-signed or invalid certificates from the mail server                                                                                                            | true                  |
| TR_MAIL_USER                 | The auth user for the mail server                                                                                                                                                      | not set               |
| TR_MAIL_PASSWORD             | The auth password for the mail server                                                                                                                                                  | not set               |
| TR_IMPORT_MAX_NESTED_LEVELS  | Max number of nested levels in terms allowed for import formats                                                                                                                        | 100                   |
| TR_AUTH_GOOGLE_ENABLED       | Set to 'true' if you'd like to enable sign-in with Google                                                                                                                              | not set               |
| TR_AUTH_GOOGLE_CLIENT_ID     | Your Google OAuth client ID                                                                                                                                                            | not set               |
| TR_AUTH_GOOGLE_CLIENT_SECRET | Your Google OAuth client secret                                                                                                                                                        | not set               |
| TR_AUTH_GOOGLE_REDIRECT_URL  | The redirect url, should be 'https://$YOUR_TRADUORA_HOST/auth/callback' and also be set in the allowed redirect URI's in the Google API console when you created the OAuth integration | not set               |


docs/changelog.md
---
id: changelog
title: Changelog
sidebar_label: Changelog
---

To download the source code for each release, check out [GitHub](https://github.com/ever-co/ever-traduora/releases). Alternatively docker images are available for each release on [Docker Hub](https://cloud.docker.com/u/everco/repository/docker/everco/ever-traduora).

## next

- TODO

## 0.19.4

- Multiple packages updates and tiny fixes
- [PR-339:](https://github.com/ever-co/ever-traduora/pull/339) Add fallback locale on export (thanks @kidoncio)
- [PR-340:](https://github.com/ever-co/ever-traduora/pull/340) Add Context on Term (thanks @kidoncio)
- [PR-310:](https://github.com/ever-co/ever-traduora/pull/310) Make the DB type as variable (thanks @azlux)

## 0.19.3

- We fixed all e2e and Unit tests in this release :rocket:
- [PR-296:](https://github.com/ever-co/ever-traduora/pull/296) Fix e2e testing
- [PR-295:](https://github.com/ever-co/ever-traduora/pull/295) More CSV Injection related tests
- [PR-294:](https://github.com/ever-co/ever-traduora/pull/294) and [PR-293:](https://github.com/ever-co/ever-traduora/pull/293) Fix some Unit Tests

## 0.19.2

- [PR-289:](https://github.com/ever-co/ever-traduora/pull/289) CSV Injection Vulnerability Fixes
- some minor packages updates (mostly for security)
- fix CircleCI builds and split e2e and unit tests running

## 0.19.1

- [PR-216:](https://github.com/ever-co/ever-traduora/pull/253) Update Angular / NestJs and other packages to latest

## 0.19.0

- [Commit](https://github.com/ever-co/ever-traduora/commit/7173276f9c049f4a8db7e7bf47fa27667421ab2c) Docker / Docker Compose improvements
- [PR-222:](https://github.com/ever-co/ever-traduora/pull/222) Support for Resource ResX files added (thanks @rai69)
- [PR-202:](https://github.com/ever-co/ever-traduora/pull/202) Add support for Docker Secrets (thanks @ynuyasha)

## 0.18.0

- [PR-216:](https://github.com/ever-co/ever-traduora/pull/216) Update Angular / NestJs and other packages to latest
- [PR-173:](https://github.com/ever-co/ever-traduora/pull/173) Security Fix for Username Enumeration (thanks @JamieSlome / @huntr-helper)

## 0.17.0

- [PR-91:](https://github.com/ever-co/ever-traduora/pull/91) Add support for PHP import/export format.
- [PR-153:](https://github.com/ever-co/ever-traduora/pull/153) Add option to debug emails locally
- [PR-122:](https://github.com/ever-co/ever-traduora/pull/122) Upgraded to Angular 9+.
- [PR-154:](https://github.com/ever-co/ever-traduora/pull/154) Add option to reject self-signed certs from mail server.
- Upgrade Docker images to Node 12.18.
- Upgraded dependencies.
- Minor performance improvements and bug-fixes.

## 0.16.0

- Allow config of max nested levels on import formats (with new default of 100).
- Pretty print JSON based exporters.

## 0.15.1

- Fix parsing of gettext (po) files with message contexts.

## 0.15.0

- [PR-72](https://github.com/ever-co/ever-traduora/pull/72) You can now organize your terms and translations with labels.
- [PR-89:](https://github.com/ever-co/ever-traduora/pull/89) Added translation progress stats per locale.
- [PR-94](https://github.com/ever-co/ever-traduora/pull/94) Fix escaping of special characters on Android Resources.
- Automatically detect file import format.
- Increase max size for file import (1MB).
- Improve error message when importing files which are too large.

## 0.14.0

- Minor bug fixes and UI improvements.
- Changed DB tables collation to support case-sensitive terms.
- [PR-87:](https://github.com/ever-co/ever-traduora/pull/87) We now support the Android Resources (XML) format.

## 0.13.0

- [PR-66:](https://github.com/ever-co/ever-traduora/pull/66) You can now invite non-traduora users into your project with the new invite system.
- Updated API reference docs.

## 0.12.1

- Ensure swagger version is same as in package.json
- expose version in `/health` endpoint

## 0.12.0

- You can now sign-in with Google.
- Fix search component case sensitivity.
- Various dependency upgrades and security fixes.

## 0.11.0

- Add swagger API docs
- Conform auth token endpoint to oauth2 conventions

## 0.10.0

- Support Apple .strings format
- Set default MySQL charset to utf8mb4. You can now use emojis in terms and translations
- Use lexical order for terms on export and find all endpoint

## 0.9.0

- Added support to import and export Gettext (po) files
- Change default encoding for Java properties format to ISO-8859-1
- Add guard against account deletion when is last project admin of project with team
- Upgrade webapp dependencies
- Minor improvements and bug fixes

## 0.8.7

- Initial open source release


api/src/config.ts
import { TypeOrmModuleOptions } from '@nestjs/typeorm';
import { DefaultNamingStrategy } from 'typeorm';

import { join } from 'path';
import * as process from 'process';

import { SnakeNamingStrategy } from './utils/snake-naming-strategy';

const env = process.env;

const getBoolOrDefault = (value: string, defaultValue: boolean) => (value ? value === 'true' : defaultValue);

export const config = {
  env: env.NODE_ENV || 'dev',
  port: parseInt(env.TR_PORT, 10) || 8080,
  secret: env.TR_SECRET || 'secret',
  virtualHost: env.TR_VIRTUAL_HOST || 'http://localhost:8080',
  publicDir: env.TR_PUBLIC_DIR || join(__dirname, '../public'),
  templatesDir: env.TR_TEMPLATES_DIR || join(__dirname, '../src/templates'),
  corsEnabled: getBoolOrDefault(env.TR_CORS_ENABLED, false),
  accessLogsEnabled: getBoolOrDefault(env.TR_ACCESS_LOGS_ENABLED, true),
  authTokenExpires: parseInt(env.TR_AUTH_TOKEN_EXPIRES, 10) || 86400,
  signupsEnabled: getBoolOrDefault(env.TR_SIGNUPS_ENABLED, true),
  maxProjectsPerUser: parseInt(env.TR_MAX_PROJECTS_PER_USER, 10) || 100,
  defaultProjectPlan: env.TR_DEFAULT_PROJECT_PLAN || 'open-source',
  autoMigrate: getBoolOrDefault(env.TR_DB_AUTOMIGRATE, true),
  import: {
    maxNestedLevels: parseInt(env.TR_IMPORT_MAX_NESTED_LEVELS, 10) || 100,
  },
  providers: {
    google: {
      active: env.TR_AUTH_GOOGLE_ENABLED,
      clientSecret: env.TR_AUTH_GOOGLE_CLIENT_SECRET,
      clientId: env.TR_AUTH_GOOGLE_CLIENT_ID,
      redirectUrl: env.TR_AUTH_GOOGLE_REDIRECT_URL,
      apiUrl: 'https://www.googleapis.com/oauth2/v4/token',
      url: 'https://accounts.google.com/o/oauth2/v2/auth',
    },
  },
  mail: {
    debug: getBoolOrDefault(env.TR_MAIL_DEBUG, false),
    sender: env.TR_MAIL_SENDER,
    host: env.TR_MAIL_HOST,
    port: parseInt(env.TR_MAIL_PORT, 10) || 587,
    secure: getBoolOrDefault(env.TR_MAIL_SECURE, false),
    rejectSelfSigned: getBoolOrDefault(env.TR_MAIL_REJECT_SELF_SIGNED, true),
    auth: {
      user: env.TR_MAIL_USER,
      pass: env.TR_MAIL_PASSWORD,
    },
  },
  db: {
    default: {
      type: env.TR_DB_TYPE || 'mysql',
      host: env.TR_DB_HOST || '127.0.0.1',
      port: parseInt(env.TR_DB_PORT, 10) || 3306,
      username: env.TR_DB_USER || 'root',
      password: env.TR_DB_PASSWORD || '',
      database: env.TR_DB_DATABASE || 'tr_dev',
      // forcing typeorm to use (mysql2) if both (mysql/mysql2) packages found, it prioritize to load (mysql)
      connectorPackage: 'mysql2',
      charset: 'utf8mb4',
      synchronize: false,
      logging: false,
      keepConnectionAlive: true,
      entities: ['src/entity/*.entity*'],
      migrations: ['src/migrations/*'],
      namingStrategy: env.TR_DB_TYPE === 'postgres' ? new SnakeNamingStrategy() : new DefaultNamingStrategy(),
    } as TypeOrmModuleOptions,
  },
};
